pkg/chartutil/dependencies.go

```diff
@@ -102,8 +102,15 @@ func getAliasDependency(charts []*chart.Chart, dep *chart.Dependency) *chart.Cha
 			continue
 		}
 
-		out := *c
 		md := *c.Metadata
+		deps := make([]*chart.Dependency, 0, len(md.Dependencies))
+		for _, dep := range md.Dependencies {
+			dep := *dep
```

Comment:
Why is this de-referencing the address, then re-assigning a new pointer? Wouldn't a `copy` be safer?

What happens when you provide a list of `nil` dependencies? Wouldn't that result in a stack trace when you attempt to de-reference `nil`? How should this code handle nil dependencies?

Reply:
A copy would not give us new Dependency structs.
We need new ones because their name are modified here which leaks on other Charts:  https://github.com/helm/helm/blob/9fafb4ad6811afb017cc464b630be2ff8390ac63/pkg/chartutil/dependencies.go#L139-L146

When you call `getAliasDependency(c.Dependencies(), req)` multiple time with the same `req.Name`, you get new Charts structs each time but their metadata.Dependencies are the same slice which come from the Chart in `c.Dependencies()` that match the Dependency (`req`).
Then we iterate over these dependencies (*chart.Chart) and call `processDependencyEnabled()`
https://github.com/helm/helm/blob/9fafb4ad6811afb017cc464b630be2ff8390ac63/pkg/chartutil/dependencies.go#L186-L191
At this point the dependencies generated by `getAliasDependency` are now the pointer `c`.
Then we iterate over `c.Metadata.Dependencies` (and call again `getAliasDependency`) then replace `req.Name` with `req.Alias`. But `c.Metadata.Dependencies` is the slice return by `getAliasDependency` in the parent, which may be shared with another Chart as seen before. 
When the `processDependencyEnabled` is called with such a Chart (same name but different alias), Dependency structs in `c.Metadata.Dependencies` have already their name changed to the alias and the following match do not work as expected:
https://github.com/helm/helm/blob/9fafb4ad6811afb017cc464b630be2ff8390ac63/pkg/chartutil/dependencies.go#L130-L133
The only Chart dependencies that are kept are the one in `c.Dependencies`, not in `c.Metadata.Dependencies`.

With a new slice and new structs for `.metadata.Dependencies`, the `req.Name = req.Alias` do not leaks on other Charts.

Not sure if I'm clear enough here, sorry :confused:.

--- 
Charts.Dependencies are apparently not supposed to contains nil pointers as this line would panic on the `req.Name` in this case: https://github.com/helm/helm/blob/9fafb4ad6811afb017cc464b630be2ff8390ac63/pkg/chartutil/dependencies.go#L130-L132
But I can add checks for nil if you prefer.